### ISSUE TITLE
Project fields or types?

### ISSUE THREAD

**Issue by clipperhouse** [clipperhouse/gen#8]
Projections present a bit of an issue in gen, which we’ve addressed with [custom methods](http://clipperhouse.github.io/gen/#Custom). These work well in allowing projection for specific member fields of a struct.

But now I am thinking that projecting specific members by name is not the future, and that projecting by _type_ is better.

For one, projection by field can feel very redundant. If your struct has several string fields, projecting each of those results in nearly-identical methods.

Secondly, one can only project on one field at a time. What if we have a FirstName and LastName field and want to ad-hoc project on a full name, which concatenates them?

Thirdly, it only works for structs! gen wants to expand to other types.

Fourthly, it’s not idiomatic gen. gen is about passing in ad-hoc funcs to get want you want, offering functionality to a _type_. The above ain’t that.

Have a look at [what we do now](http://clipperhouse.github.io/gen/#Custom), and compare to:

```
// gen:"Where,GroupBy,Select" project:"string,OtherType"
type Thing struct {
    Name string
    Country string
    Score int
}
```

...which would result in:

```
func (rcv Things) SelectString(func (*Thing) string) []string
func (rcv Things) GroupByString(func (*Thing) string) map[string]Things

func (rcv Things) SelectOtherType(func (*Thing) OtherType) []OtherType
func (rcv Things) GroupByOtherType(func (*Thing) OtherType) map[OtherType]Things
```

In a nutshell, field tags (specifying methods) go away, replaced with a new `project` tag (specifying types). All methods move to the `gen` tag. So, grouping by country would be:

```
tisOfThee := myThings.GroupByString(func (t *Thing) string {
    return t.Country
})
```

Thoughts?
